Comments: new CommentLikes component

rodrigoi opened the pull request:
This PR creates a new `CommentLikes` block component, connected to the store but with no action dispatches.
It accepts the like and unlike actions as props, and it shows a like count.

How to test:

One the comment management interface, like or unlike a comment. The new like status should persist.
The button should show the liked status only if the current user liked the comment. The comment count should not be displayed if the only like is from the current user, and the label should display the _Liked_ legend. Otherwise, the like count should appear next to a _like/likes_ legend.

Closes #13910

gwwar commented on client/blocks/comment-likes/index.jsx:
```diff
@@ -0,0 +1,67 @@
+			onClick={ i_like ? dispatchUnlike : dispatchLike }
+		>
+			<Gridicon icon={ i_like ? 'star' : 'star-outline' } />
+			<span>{ showLikeCount ? like_count + ' ' : '' }{ likeLabel }</span>
```
I think https://github.com/Automattic/i18n-calypso/#pluralization might be a better fit. It might be the case that certain languages would inverse the count/label here.

rodrigoi replied:
A better fit indeed. Updated on 109e2ee3440135bc2fdfa2c67b8acf314e72baeb.

dmsnell commented on client/blocks/comment-likes/index.jsx:
```diff
@@ -15,19 +15,20 @@ import { likeComment, unlikeComment } from 'state/comments/actions';
 
 const getLabel = ( i_like, like_count, translate ) => {
+	if ( i_like && like_count === 1 ) {
+		return translate( 'Liked', { comment: 'Displayed when a person "likes" a post.' } );
```
I know that the comments are generally safe here in the translations and I wonder if we need it.

On the other hand, this would definitely make it easier for people to translate gender, since they will presumably know what gender a post is.

@yoavf this might be a good example of a situation like we spoke about. do you have any insightful comment here?

Copons commented on client/blocks/comment-detail/comment-detail-actions.jsx:
```diff
@@ -38,20 +40,7 @@ export const CommentDetailActions = ( {
-					}</span>
-				</Button>
-			}
+			{ hasAction( commentStatus, 'like' ) && <CommentLikes { ...{ siteId, postId, commentId } } />}
```
Missing space before the last `}`. Not sure why eslint isn't catching that though. 🤔 

Also, but this is 100% personal preference, I usually go multiline with conditionally rendered components:
```es6
{ hasAction( commentStatus, 'like' ) &&
	<CommentLikes { ...{ siteId, postId, commentId } } />
}
```
(But it doesn't really matter anyway)

Copons commented on client/blocks/comment-likes/index.jsx:
```diff
@@ -0,0 +1,67 @@
+		return translate( 'Like', {
+			context: 'verb: imperative',
+			comment: 'Label for a button to "like" a post.',
+		} );
```
Awesome use of i18n!
Note to self: do like @rodrigoi does!

Copons commented on client/blocks/comment-likes/index.jsx:
```diff
@@ -0,0 +1,67 @@
+	translate,
+} ) => {
+	const { i_like, like_count } = commentLike;
+	const likeLabel = getLabel( i_like, like_count, translate );
```
We could spare the constant and render `getLabel()` directly.

Copons commented on client/blocks/comment-detail/comment-detail-actions.jsx:
```diff
@@ -22,12 +23,13 @@ const commentActions = {
 export const CommentDetailActions = ( {
+	siteId,
+	postId,
+	commentId,
```
I always alphabetize props, destructuring, etc.
On the other hand, I find clearer to have IDs on top (also ordered after the actions/selectors arguments), so yeah, another 100% personal preference. :)

dmsnell replied:
I typically prefer this ordering to alphabetical when it's so clearly established in an existing hierarchy. That's my vote at least.

gwwar commented:
> we're losing the "Comment approved" notice on "like and approve"

@Copons it seems like we should be able to resolve that by moving logic to the notices middleware, or by adding an onClick prop to the new block. Let's let this land for now and we can add another issue to fix the notice/undo

rodrigoi commented:
@Copons @gwwar I've added a `successNotice` dispatch as a side effect under a feature flag on the data-layer. This is the commit: 414a0e09e5844e2d0991430ff55e54ebef4115cd for the side effect, and this is the commit for the feature flag: 97b05561c637a7c73c278f41eaae8b3dd3437521.

I've added a feature flag to prevent the notice to open when users like comments on the reader. We can eventually add a `meta` property to the action to prevent the reader to show the notice.

dmsnell commented on client/blocks/comment-likes/index.jsx:
```diff
@@ -0,0 +1,66 @@
+
+const mapDispatchToProps = ( dispatch, { siteId, postId, commentId, commentStatus } ) => ( {
+	dispatchLike: () => dispatch( likeComment( siteId, postId, commentId, commentStatus ) ),
+	dispatchUnlike: () => dispatch( unlikeComment( siteId, postId, commentId ) )
```
I'll note here that I don't think the React component cares at all that these are dispatching likes or unlikes. To the component I think we should spec out an API that's meaningful to the behaviors it intends on triggering. In this case, just `likeComment()` and `unlikeComment()` or something like that. That it happens via a Redux store dispatch should be meaningless to the component. Otherwise, we will have to build our terminology and tests around Redux when that point is completely unrelated to the UI we're building.

dmsnell commented on client/state/data-layer/wpcom/sites/comments/likes/new/index.js:
```diff
@@ -23,13 +25,29 @@ export const likeComment = ( { dispatch }, action ) => {
+			id: noticeId,
+			isPersistent: true,
+			button: translate( 'Undo' ),
+			onClick: () => dispatch( changeCommentStatus( siteId, postId, commentId, 'unapproved' ) )
```
This is confusing me here. It looks like we're presenting the _like_ action an undo which sets a comment as `unapproved`. Aren't we mixing behaviors? What if the comment was formerly approved, it was liked, then that like was undone? I think it's not actually undoing but performing a different change to the comment.

Copons replied:
Just pointing out that, in the like context, the undo is only for "like and approve" actions, and performs an "unlike and unapprove" action.

dmsnell replied:
why?

(also, it doesn't appear to undo the like)

Copons replied:
Because the like button is a toggle. If you want to undo a like, well, just click on it again. 🙂 

All the other actions, instead, move comments in other lists, making them disappear from the current view (mind that this does not happen with the CommentFaker).
So without an undo, if you accidentally approve a comment, you'd need to go to the approved list and hunt down the comment, which might not be in the first page.

At the same time, we don't really need a like/unlike notice. The behaviour is, in my opinion, well established everywhere. We might not use the thumbs-up or the heart icon, but its pretty clear what it does (of course I'll be disproved by the next user test on this :D ).

More importantly, like/unlike, by itself, is a "non-destructive" action (lack of better term). No matter its like status, the comment will remain in its place.
Though, this is not the case when liking unapproved comments. There, the comment is also moved to the approved list (which means: it shows up to all users).
As the approve action is "destructive", it has its own notice and undo. So we give to the "like+approve" action the same notice and undo (which of course should also undo the like; I guess my bad, somehow I thought unapproving also automatically unliked).

dmsnell replied:
To me this is revealing a bigger problem that might need to be addressed: that as soon as you like the comment it disappears (wholly unexpected!)

Is an _undo_ notice the most appropriate response to this problem? What other options could we explore?

For example, what if we kept a comment in the unapproved list after liking it, somehow indicating that now it's approved, but not switching lists? We do this in the notifications client with unread notifications, for example. The background color can indicate that it's not approved, plus the change of buttons. Once we navigate to another list we can stop faking it.

@drw158 are you familiar with this undo behavior?

drw158 replied:
The plan is/was to keep the comment around until page refresh or filter change. So, the user would approve a comment, and it would not disappear. If the user changes tabs or reloads the page, then the comment will be in its proper place. This behavior is the same as the Unread filter in Mail.app on iOS and others (like our own notifications).

We just haven't gotten that far in the project yet. Probably will need to be an m2 thing. I actually just noticed that there is not an issue for it, sorry about that. I think we've talked about it in Slack, but just haven't documented it.

I think it's still valuable to have a notice when a user performs an action. Including the Undo button will instill confidence that they can revert an action if they ever do it accidentally. They can of course revert without the Undo button as well.

drw158 replied:
No notice or undo is needed for Like.

drw158 replied:
If a users likes a comment, it should approve the comment as well. As a result, a notice should be shown, saying that the comment was approved, along with the undo button. The undo button would only revert the approve, not the like.

Copons replied:
I think the approved comment disappearance is my fault actually.
To be fair though, I'd say that implementing the logic that prevents approved comments to disappear requires the actual comments data pulled from the Redux state, because it isn't easily fakeable with the CommentFaker.

Also, just to be clear: this "persistence" behaviour is only applied to "unapproved -> approved" or to any other action that moves comments between lists (e.g. "spam -> approved", etc.)?

About the whole "like+approve": since a "liked unapproved" comment doesn't really make sense, I've always worked with the impression that "undo like+approve" = "unlike+unapprove".

dmsnell replied:
> Also, just to be clear: this "persistence" behaviour is only applied to "unapproved -> approved" or to any other action that moves comments between lists (e.g. "spam -> approved", etc.)?

confirmed :smile:

dmsnell replied:
> About the whole "like+approve": since a "liked unapproved" comment doesn't really make sense, I've always worked with the impression that "undo like+approve" = "unlike+unapprove".

I find the whole split of the actions as confusing too. I really struggle to see the connection of undoing an action I didn't knowingly trigger. It's as if I made a change to a post and hit undo that then that undo unpublished the post.

drw158 replied:
To go a bit further on the previous comment, I **do** think it would be ok if the comment becomes unliked. I also don't think "liked + unapproved" makes sense. Sorry if this is contrary to what I said. This is getting pretty confusing :)

Copons replied:
@drw158 I think that's exactly what you've been saying all along, but yeah, the terminology is definitely confusing 😄

dmsnell commented:
> I've added a feature flag to prevent the notice to open when users like comments on the reader. We can eventually add a meta property to the action to prevent the reader to show the notice.

Why would we want separate behaviors? By standardizing on Redux actions and centralizing these kinds of interactions in the data layer we have the ability to bring more consistency to the platform. Seems like if we want an undo in comments management it would make sense wherever we like comments (for example, in the notifications panel).

Copons commented:
All right, thanks @rodrigoi for the update!

I'm not sold on putting the notice logic in the data layer, but we can discuss and work it out another time. 🙂 

I've noticed three odd behaviours I'd like to point out:

- I didn't expect so much lag between the click and the result of the action, like 1 to 3 seconds. It feels weird, especially after being used to FB/Twitter where like actions are instant. I don't know if this is a technical limitation we can work around, but in case we could work in a "three-state" system for the action buttons: off / loading / on.

- When undoing a "like+approve" the comment gets unapproved but also keeps its like.

- This is the only change I'd like to request UI-wise:
Add a `removeNotice` in the [undo callback](https://github.com/Automattic/wp-calypso/blob/414a0e09e5844e2d0991430ff55e54ebef4115cd/client/state/data-layer/wpcom/sites/comments/likes/new/index.js#L47), e.g.:
```es6
onClick: () => {
	dispatch( removeNotice( noticeId ) );
	dispatch( changeCommentStatus( siteId, postId, commentId, 'unapproved' ) );
}
```
Otherwise, the "Comment Approved - Undo" notice will stay visible until its 5000ms timeout.

gwwar commented on client/state/data-layer/wpcom/sites/comments/likes/new/index.js:
```diff
@@ -23,13 +25,29 @@ export const likeComment = ( { dispatch }, action ) => {
+		like_count
+	} ) );
+
+	if ( isEnabled( 'comments/undo-in-datalayer' ) && 'unapproved' === status ) {
```
Instead of adding a feature flag here, why not add notice middleware for `COMMENTS_LIKE` we can for example check the current route and only dispatch a notice if we're in the comments management path.

gwwar replied:
But otherwise I don't see too much harm in displaying this notice for all cases. We'd just need to verify in the other areas that actually dispatch this.

gwwar commented:
Can we get unblocked here for either this PR or #16082 ? Since this is not a released feature, we're welcome to iterate on this in an incomplete fashion. Eg adding likes, breaking the undos, and addressing that in a follow up PR.

dmsnell commented:
I'm fine merging this. I really don't like how coupled we have gotten to the API (`i_like` for example) but that's been a losing battle from the outset.

Copons commented:
About this PR: I've proposed a different handling comment actions approach in  https://github.com/Automattic/wp-calypso/pull/16082.
I believe that's the way to go, but there are plenty of good parts I would like to keep from this PR.
This is my reasoning as explained in an internal conversation.

---

To be clear on what’s happening: right now we wrap all the things with the so called **`CommentFaker`**, that loads the comments into its own state (so that any actions won't be permanent) and pass them, along with fake actions, to the actual **`CommentList`** as props.
Then, `CommentList` acts as a centralized controller that takes care of handling the actions, their notices, and their undos (because we have both single and bulk actions).

The idea was to simply kill the Faker and replace `props.comments`, `props.setStatus()`, etc. with real Redux selectors and actions.
In fact it's just slightly more complicated: we need to also adjust some function names, maybe some parameters and some checks, but nothing major.

---

This PR takes care of the **like** action, which is incidentally the most annoying one, because it also has the **approve** side-effect.

He creates a new `CommentLikes` component that is a multistate button:
- “Like” when there are no likes.
- “Liked (by you)” when there’s only your like.
- “n Likes” if there are more than 1 likes (and it can be highlighted if you liked it).

This component is indeed needed and I’d have no problems with it, except… it’s a **block**, which means it’s connected, which means it decentralizes the actions away from `CommentList`.
Normally it would be a good thing, but again, we also have bulk actions to handle here and we still need a centralized controller anyway!

Also, `CommentLikes` dispatches the Redux actions directly. This means two things:
- The `likeComment()` action and API call need a new `status` parameter to approve the comment (a good thing, as it spares us the need to dispatch two actions, one for the like and one for the approve).
- We end up introducing the notices logic inside the data layer (and this imho is not a good thing).

What if for some reason we want to dispatch comment actions without the notices?
What about the fact that the notices are presentational, and not related to the data?
Last but not least, we share the comment actions with the Reader: is it worth it introducing side effects to the Reader?

---

[My PR](https://github.com/Automattic/wp-calypso/pull/16082) instead takes care of **all** the actions.
It simply replaces `CommentFaker.setStatus()` and `CommentFaker.toggleLike()` with (let’s say) `Redux.setStatus()` and `Redux.likeComment()` and `Redux.unlikeComment()`.

The entire notices logic stays inside the `CommentList`, where we have full control of everything. We can do and undo stuff as much as we want and nothing else in Calypso would even _notice_ it (pun unintended).

---

Now, the thing here is that in this PR there are plenty of good stuff:
- The work on the `CommentLikes` component is legit awesome and definitely something I’ve never considered.
- Adding the approve side effect directly to the Redux actions and the API calls is :+1: (I was somehow sure it was already working like that :smile:).

And some things that I don’t agree with:
- Moving the actions to the button components.
- Handling notices inside the data layer.

---

So, the idea here is pretty much this:
- Remove the notices side effects from the data layer.
- Dis-`connect` the `CommentLikes` component, and making it controlled from the outside.

Copons commented:
@dmsnell, @gwwar: I've updated this PR removing the dispatches from the `CommentLikes` block and the notices from the data layer.

kwight commented:
Took a look at this; tests great for me, including without a connection (reverts to "unliked" status, gives notice).

> Dis-connect the CommentLikes component, and making it controlled from the outside.

@Copons does this mean there's still more to do on this? I can see the component is still `connect`ed, but it makes sense.

Also curious what the relationship is of this component to `client/blocks/comments/comment-likes.jsx` – seems like we're replacing that one, can it be removed? It also has some stats code, do we need to add that into this new component?

Copons commented:
@kwight 

> Dis-connect the CommentLikes component, and making it controlled from the outside.

This was an overlook: I've only removed the dispatches, but I still need to keep the component connected for the number of likes.

> Also curious what the relationship is of this component to `client/blocks/comments/comment-likes.jsx` – seems like we're replacing that one, can it be removed?

Good question: I'm not sure what plans @rodrigoi had about it. Should look deeper.

> It also has some stats code, do we need to add that into this new component?

I always forget about stats and we should add them everywhere in Comments Management, but near the actions I guess (so, in this case in `CommentList`).

gwwar commented:
👍  I think this should be good to ship after we add a Readme that explains the use case, and why we'd want to connect to see the count but not handle the dispatch automatically. Also let's update the PR summary so we're not confused when we look back at this.

Copons commented:
Closing this PR because it turned out we don't really want the like count to be shown in the Comment Management.
